README.md
<div align="center">
  <img src="https://github.com/nrdlab/pattern-matching-ts/blob/master/img/pattern-matching-ts.png?raw=true">
</div>

<h4 align="center">
  <strong>Pattern matching</strong> in <strong>Typescript</strong>.
</h4>

<p>
  <strong>Pattern Matching</strong> is a declarative much more powerful and less verbose alternative to imperatives <i>"if/else"</i> conditions.<br/>
  A definition can be found inside <a href="https://docs.scala-lang.org/tour/pattern-matching.html"><strong>Scala Documentation</strong></a><br><br>
<i>“Pattern matching tests whether a given value (or sequence of values) has the shape defined by a pattern, and, if it does, binds the variables in the pattern to the corresponding components of the value (or sequence of values).”</i><br><br>
  In <strong>Functional Programming languages</strong>, there're built-in keywords for <strong>Pattern Matching</strong>. <strong>Typescript</strong> though is one language that works very well with <strong>Functional Programming</strong> but lacks this feature.<br/>
This package aims to bring <strong>Pattern Matching</strong> feature to <strong>Typescript</strong> through <strong>Discriminated Union Types</strong> / <strong>Algebraic Data Types</strong>.
</p>

<p align="center">
  <img alt="npm" src="https://img.shields.io/npm/v/pattern-matching-ts?color=green&logo=yarn&logoColor=white&style=flat-square">
      <a href="https://codecov.io/gh/nrdlab/pattern-matching-ts">
        <img src="https://codecov.io/gh/nrdlab/pattern-matching-ts/branch/master/graph/badge.svg?token=1V23E6VDHN"/>
      </a>
<a href="https://github.com/nrdlab/pattern-matching-ts/blob/master/LICENSE"><img alt="GitHub license" src="https://img.shields.io/github/license/nrdlab/pattern-matching-ts?style=flat-square"></a>
<a href="https://github.com/nrdlab/pattern-matching-ts/stargazers"><img alt="GitHub stars" src="https://img.shields.io/github/stars/nrdlab/pattern-matching-ts?color=green&style=flat-square"></a>
<a href="https://github.com/nrdlab/pattern-matching-ts/network"><img alt="GitHub forks" src="https://img.shields.io/github/forks/nrdlab/pattern-matching-ts?color=green&style=flat-square"></a>
  <a href="https://github.com/nrdlab/pattern-matching-ts/actions?query=workflow%3ACI">
  <img alt="GitHub Workflow Status" src="https://img.shields.io/github/workflow/status/nrdlab/pattern-matching-ts/CI?style=flat-square&color=green">
  <a>
</p>

<br />

# Index

- [Installation](#installation)
- [Usage](#usage)

---

- [Match](#match)
  - [Match Option](#option-match)
  - [Match Default ](#default-match)
- [MatchW](#matchW)
  - [MatchW Option](#option-MatchW)
  - [MatchW Either](#either-MatchW)
  - [MatchW Default](#default-matchW)

---

- [License](#license)

# Installation

**yarn**

```sh
yarn add pattern-matching-ts
```

**npm**

```sh
npm install --save pattern-matching-ts
```

# Usage

**_MatchW_**

> ### Option MatchW

```ts
import * as M from 'pattern-matching-ts/lib/match'
import { pipe } from 'fp-ts/lib/function'
import * as O from 'fp-ts/lib/Option'

const optionMatching = (o: unknown) =>
  pipe(
    o,
    M.matchW('_tag')({
      Some: ({ value }) => 'Something: ' + value,
      None: () => 'Nothing',
      _: () => 'Default'
    })
  )

assert.deepStrictEqual(optionMatching(O.some('data')), 'Something: data')
assert.deepStrictEqual(optionMatching(O.none), 'Nothing')
assert.deepStrictEqual(optionMatching((undefined as unknown) as O.None), 'Default')
```

> ### Either MatchW

```ts
import * as M from 'pattern-matching-ts/lib/match'
import { pipe } from 'fp-ts/lib/function'
import * as E from 'fp-ts/lib/Either'

type RGB = Record<'r' | 'g' | 'b', number>
const either = (maybeRgb: E.Either<string, RGB>) =>
  pipe(
    maybeRgb,
    M.matchW('_tag')({
      Left: ({ left }) => 'Error: ' + left,
      Right: ({ right: { r, g, b } }) => `Red: ${r} | Green: ${g} | Blue: ${b}`
    })
  )

assert.deepStrictEqual(either(E.right({ r: 255, g: 255, b: 0 })), 'Red: 255 | Green: 255 | Blue: 0')
```

> ### Default MatchW

```ts
import * as M from 'pattern-matching-ts/lib/match'
import { pipe } from 'fp-ts/lib/function'

interface ServerResponse<Code extends string | number> {
  readonly code: Code
}

interface Response<Body> {
  readonly response: {
    readonly body: Body
  }
}

interface Success extends ServerResponse<200>, Response<ReadonlyArray<string>> {}

interface NotFoundError extends ServerResponse<404> {}

interface ServerError extends ServerResponse<500> {
  readonly detail: string
}

type Responses = Success | NotFoundError | ServerError

const matchResponse = (response: Responses) =>
  pipe(
    response,
    M.matchW('code')({
      500: ({ detail }) => ({ message: 'Internal server error', detail }),
      404: () => ({ message: 'The page cannot be found!' }),
      200: ({ response }) => response.body,
      _: () => 'Unexpected response'
    })
  )

assert.deepStrictEqual(either(E.right({ r: 255, g: 255, b: 0 })), 'Red: 255 | Green: 255 | Blue: 0')
assert.deepStrictEqual(matchResponse({ code: 200, response: { body: ['data'] } }), ['data'])
assert.deepStrictEqual(matchResponse({ code: 500, detail: 'Cannot connect to the database' }), {
  message: 'Internal server error',
  detail: 'Cannot connect to the database'
})
assert.deepStrictEqual(matchResponse({ code: 404 }), { message: 'The page cannot be found!' })
```

**_Match_**

> ### Option Match 

```ts
import * as M from 'pattern-matching-ts/lib/match'
import * as O from 'fp-ts/lib/Option'

const optionMatching = M.match<O.Option<string>, string>({
  Some: (x) => `Something: ${x.value}`,
  None: () => 'Nothing'
})

assert.deepStrictEqual(optionMatching(O.some('data')), 'Something: data')
assert.deepStrictEqual(optionMatching(O.none), 'Nothing')
```

> ### Default Match

```ts
import * as M from 'pattern-matching-ts/lib/match'

interface ChangeColor<T = number> {
  readonly _tag: 'ChangeColor'
  readonly value: {
    readonly r: T
    readonly g: T
    readonly b: T
  }
}
interface Move<T = number> {
  readonly _tag: 'Move'
  readonly value: {
    readonly x: T
    readonly y: T
  }
}
interface Write {
  readonly _tag: 'Write'
  readonly value: {
    readonly text: string
  }
}

type Cases = ChangeColor<number> | Move | Write
const matchMessage = M.match<Cases, string>({
  ChangeColor: ({ value: { r, g, b } }) => `Change the color to Red: ${r} | Green: ${g} | Blue: ${b}`,
  Move: ({ value: { x, y } }) => `Move in the x direction: ${x} and in the y direction: ${y}`,
  Write: ({ value: { text } }) => `Text message: ${text}`,
  _: () => 'Default message'
})

const ChangeColor = ({ r, g, b }: ChangeColor<number>['value']): ChangeColor<number> => ({
  _tag: 'ChangeColor',
  value: { r, g, b }
})

const Move = ({ x, y }: Move['value']): Move => ({
  _tag: 'Move',
  value: { x, y }
})

const Write = ({ text }: Write['value']): Write => ({
  _tag: 'Write',
  value: { text }
})

assert.deepStrictEqual(
  matchMessage(Move({ x: 500, y: 100 })),
  'Move in the x direction: 500 and in the y direction: 100'
)

assert.deepStrictEqual(
  matchMessage(ChangeColor({ r: 12, g: 20, b: 30 })),
  'Change the color to Red: 12 | Green: 20 | Blue: 30'
)

assert.deepStrictEqual(matchMessage(Write({ text: 'my message' })), 'Text message: my message')
```

Blog posts that introduces the API.

 👉 [Pattern Matching in Typescript](https://dev.to/stefano_regosa/typescript-pattern-matching-ne8)

 👉 [Pipeable Pattern Matching in Typescript](https://dev.to/stefano_regosa/pipeable-pattern-matching-in-typescript-33dn)


[MIT](/LICENSE.md)


CHANGELOG.md
# 2.0.0

- **New Feature**
  - add `matchW` module


test/match.test.ts
import * as assert from 'assert'
import * as O from 'fp-ts/lib/Option'
import * as M from '../src/match'

interface ChangeColor<T = number> {
  readonly _tag: 'ChangeColor'
  readonly value: {
    readonly r: T
    readonly g: T
    readonly b: T
  }
}
interface Move<T = number> {
  readonly _tag: 'Move'
  readonly value: {
    readonly x: T
    readonly y: T
  }
}
interface Write {
  readonly _tag: 'Write'
  readonly value: {
    readonly text: string
  }
}

describe('pattern matching', () => {
  const optionMatching = M.match<O.Option<string>, string>({
    Some: (x) => `Something: ${x.value}`,
    None: () => 'Nothing'
  })

  type Cases = ChangeColor<number> | Move | Write
  const matchMessage = M.match<Cases, string>({
    ChangeColor: ({ value: { r, g, b } }) => `Change the color to Red: ${r} | Green: ${g} | Blue: ${b}`,
    Move: ({ value: { x, y } }) => `Move in the x direction: ${x} and in the y direction: ${y}`,
    Write: ({ value: { text } }) => `Text message: ${text}`,
    _: () => 'Default message'
  })

  it('Option', () => {
    assert.deepStrictEqual(optionMatching(O.some('data')), 'Something: data')
    assert.deepStrictEqual(optionMatching(O.none), 'Nothing')
  })

  it('match', () => {
    const ChangeColor = ({ r, g, b }: ChangeColor<number>['value']): ChangeColor<number> => ({
      _tag: 'ChangeColor',
      value: { r, g, b }
    })
    assert.deepStrictEqual(
      matchMessage(ChangeColor({ r: 12, g: 20, b: 30 })),
      'Change the color to Red: 12 | Green: 20 | Blue: 30'
    )

    const Move = ({ x, y }: Move['value']): Move => ({
      _tag: 'Move',
      value: { x, y }
    })
    assert.deepStrictEqual(
      matchMessage(Move({ x: 500, y: 100 })),
      'Move in the x direction: 500 and in the y direction: 100'
    )

    const Write = ({ text }: Write['value']): Write => ({
      _tag: 'Write',
      value: { text }
    })
    assert.deepStrictEqual(matchMessage(Write({ text: 'my message' })), 'Text message: my message')
  })

  it('match default', () => {
    assert.deepStrictEqual(matchMessage((null as unknown) as ChangeColor), 'Default message')
    assert.deepStrictEqual(matchMessage((undefined as unknown) as ChangeColor), 'Default message')
  })
})


test/matchW.test.ts
import assert from 'assert'
import { pipe } from 'fp-ts/lib/function'
import * as O from 'fp-ts/lib/Option'
import * as E from 'fp-ts/lib/Either'
import * as M from '../src/match'

describe('pattern matching pipeble', () => {
  interface ServerResponse<Code extends string | number> {
    readonly code: Code
  }

  interface Response<Body> {
    readonly response: {
      readonly body: Body
    }
  }
  interface Success extends ServerResponse<200>, Response<ReadonlyArray<string>> {}

  interface NotFoundError extends ServerResponse<404> {}
  interface ServerError extends ServerResponse<500> {
    readonly detail: string
  }

  type Responses = Success | NotFoundError | ServerError

  const optionMatching = (o: O.Option<string>) =>
    pipe(
      o,
      M.matchW('_tag')({
        Some: ({ value }) => 'Something: ' + value,
        None: () => 'Nothing',
        _: () => 'Default'
      })
    )

  it('Option', () => {
    assert.deepStrictEqual(optionMatching(O.some('data')), 'Something: data')
    assert.deepStrictEqual(optionMatching(O.none), 'Nothing')
    assert.deepStrictEqual(optionMatching((undefined as unknown) as O.None), 'Default')
    assert.deepStrictEqual(optionMatching((null as unknown) as O.None), 'Default')
    assert.deepStrictEqual(optionMatching(('' as unknown) as O.None), 'Default')
  })
  type RGB = Record<'r' | 'g' | 'b', number>
  const either = (maybeRgb: E.Either<string, RGB>) =>
    pipe(
      maybeRgb,
      M.matchW('_tag')({
        Left: ({ left }) => 'Error: ' + left,
        Right: ({ right: { r, g, b } }) => `Red: ${r} | Green: ${g} | Blue: ${b}`
      })
    )

  const eitherMatchingLeft = pipe(
    E.left('RGB color values not found'),
    M.matchW('_tag')({
      Right: ({ right: { r, g, b } }) => `Red: ${r} | Green: ${g} | Blue: ${b}`,
      Left: ({ left }) => 'Error: ' + left
    })
  )

  it('Either', () => {
    assert.deepStrictEqual(either(E.right({ r: 255, g: 255, b: 0 })), 'Red: 255 | Green: 255 | Blue: 0')
    assert.deepStrictEqual(eitherMatchingLeft, 'Error: RGB color values not found')
  })

  const matchResponse = (response: Responses) =>
    pipe(
      response,
      M.matchW('code')({
        500: ({ detail }) => ({ message: 'Internal server error', detail }),
        404: () => ({ message: 'The page cannot be found!' }),
        200: ({ response }) => response.body,
        _: () => 'Unexpected response'
      })
    )

  it('Response', () => {
    assert.deepStrictEqual(matchResponse({ code: 200, response: { body: ['data'] } }), ['data'])
    assert.deepStrictEqual(matchResponse({ code: 500, detail: 'Cannot connect to the database' }), {
      message: 'Internal server error',
      detail: 'Cannot connect to the database'
    })
    assert.deepStrictEqual(matchResponse({ code: 404 }), { message: 'The page cannot be found!' })
  })

  it('Default', () => {
    assert.deepStrictEqual(matchResponse((null as unknown) as ServerError), 'Unexpected response')
    assert.deepStrictEqual(matchResponse((undefined as unknown) as ServerError), 'Unexpected response')
    assert.deepStrictEqual(matchResponse(({ code: '_', value: '' } as unknown) as ServerError), 'Unexpected response')
  })
})


src/matchP.ts
import { pipe } from 'fp-ts/lib/function'
import * as O from 'fp-ts/lib/Option'
import * as E from 'fp-ts/lib/Either'

export type NonTagType<A, K extends keyof A, Type extends string> = Omit<Extract<A, { [k in K]: Type }>, K>

type Matcher<_Tag extends string, A extends { [k in _Tag]: string }, R = unknown> = {
  [K in A[_Tag]]: (v: NonTagType<A, _Tag, K>) => R
}
type ReturnFromMatcher<A extends Record<string, unknown>> = {
  [P in keyof A]: A[P] extends (...args: infer P) => infer R ? (...args: P) => R : never
}[keyof A]

interface DefaultCase { readonly '_': () => any }

export const matchP = <_Tag extends string>(
  _tag: _Tag /* _tag */
) => <A extends { [k in _Tag ]: string }>(
  matcher: Matcher<_Tag, A> | DefaultCase /* Case */
) => <R extends ReturnFromMatcher<Matcher<_Tag, A>>>(
  v: A
): R => {
  return (matcher as any)[typeof v?.[_tag] !== 'undefined' ? v?.[_tag] : '_' ](v)
}

pipe(
  O.some('wow'),
  matchP('_tag')({
    None: () => 'oh noooo ',
    Some: ({ value }) => 'some ' + value,
    _: () => 'stocazzo'
  })
)
export interface Left<E> {
  readonly _tag: 'Left'
  readonly left: E
}
export interface Right<A> {
  readonly _tag: 'Right'
  readonly right: A
}
export type Either<E, A> = Left<E> | Right<A>
export const left = <E = never, A = never>(e: E): Either<E, A> => ({ _tag: 'Left', left: e })
pipe(
  E.left({ r: 'R', g: 'G', b: 1 }),
  matchP('_tag')({
    Left: ({ left: { r, g, b } }) => `Red: ${r} | Green: ${g} | Blue: ${b}`,
    Right: ({ right }) => 'is right' + right
  })
)


src/matchW.ts
const DEFAULT = '_'
interface DefaultCase {
  readonly [DEFAULT]: () => unknown
}

type Match<A, _Tag extends string, X, R> = (p: Extract<A, { [_tag in _Tag]: X }>) => R
/**
 * A Wider pipeable Pattern Maching Implementation
 *
 * @example
 * import * as M from 'pattern-matching-ts/lib/matchW'
 *
 * const optionMatching = (o: O.Option<string>) =>
 *    pipe(
 *      o,
 *      M.matchW('_tag')({
 *       Some: ({ value }) => 'Something: ' + value,
 *       None: () => 'Nothing',
 *     }))
 *
 *   assert.deepStrictEqual(optionMatching(O.some('data')), 'Something: data')
 *   assert.deepStrictEqual(optionMatching(O.none), 'Nothing')
 *
 * const matchResponse = (response: Responses) =>
 *   pipe(
 *     response,
 *     M.matchW('code')({
 *        500: ({ detail }) => ({ message: 'Internal server error', detail }),
 *        404: () => ({ message: 'The page cannot be found!' }),
 *        200: ({ response }) => response.body,
 *        _: () => 'Unexpected response'
 *     })
 *   )
 *
 * it('Response', () => {
 *   assert.deepStrictEqual(matchResponse({ code: 200, response: { body: ['data'] } }), ['data'])
 *   assert.deepStrictEqual(matchResponse({ code: 500, detail: 'Cannot connect to the database' }), {
 *     message: 'Internal server error',
 *     detail: 'Cannot connect to the database'
 *   })
 *   assert.deepStrictEqual(matchResponse({ code: 404 }), { message: 'The page cannot be found!' })
 * })
 */
export const matchW: <_Tag extends string>(
  _tag: _Tag
) => {
  <
    A extends { [X in _Tag]: string | number | typeof DEFAULT },
    K extends { [X in A[_Tag]]: Match<A, _Tag, X, unknown> }
  >(
    k: K
  ): (match: A) => ReturnType<K[keyof K]> | DefaultCase
} = (_tag) => (k: any) => (match: any | DefaultCase) => {
  /* istanbul ignore next */
  return k?.[match?.[_tag]] ? k[match[_tag]](match) : k[DEFAULT](match)
}


scripts/build.ts
import * as path from 'path'
import * as E from 'fp-ts/Either'
import { pipe } from 'fp-ts/function'
import * as RTE from 'fp-ts/ReaderTaskEither'
import * as A from 'fp-ts/ReadonlyArray'
import * as TE from 'fp-ts/TaskEither'
import { FileSystem, fileSystem } from './FileSystem'
import { run } from './run'

interface Build<A> extends RTE.ReaderTaskEither<FileSystem, Error, A> {}

const OUTPUT_FOLDER = 'dist'
const PKG = 'package.json'

export const copyPackageJson: Build<void> = (C) =>
  pipe(
    C.readFile(PKG),
    TE.chain((s) => TE.fromEither(E.parseJSON(s, E.toError))),
    TE.map((v) => {
      const clone = Object.assign({}, v as any)

      delete clone.scripts
      delete clone.files
      delete clone.devDependencies

      return clone
    }),
    TE.chain((json) => C.writeFile(path.join(OUTPUT_FOLDER, PKG), JSON.stringify(json, null, 2)))
  )

export const FILES: ReadonlyArray<string> = ['CHANGELOG.md', 'LICENSE', 'README.md']

export const copyFiles: Build<ReadonlyArray<void>> = (C) =>
  pipe(
    FILES,
    A.traverse(TE.taskEither)((from) => C.copyFile(from, path.resolve(OUTPUT_FOLDER, from)))
  )

const traverse = A.traverse(TE.taskEither)

export const makeModules: Build<void> = (C) =>
  pipe(
    C.glob(`${OUTPUT_FOLDER}/lib/*.js`),
    TE.map(getModules),
    TE.chain(traverse(makeSingleModule(C))),
    TE.map(() => undefined)
  )

function getModules(paths: ReadonlyArray<string>): ReadonlyArray<string> {
  return paths.map((filePath) => path.basename(filePath, '.js')).filter((x) => x !== 'index')
}

function makeSingleModule(C: FileSystem): (module: string) => TE.TaskEither<Error, void> {
  return (m) =>
    pipe(
      C.mkdir(path.join(OUTPUT_FOLDER, m)),
      TE.chain(() => makePkgJson(m)),
      TE.chain((data) => C.writeFile(path.join(OUTPUT_FOLDER, m, 'package.json'), data))
    )
}

function makePkgJson(module: string): TE.TaskEither<Error, string> {
  return pipe(
    JSON.stringify(
      {
        main: `../lib/${module}.js`,
        module: `../es6/${module}.js`,
        typings: module === 'HKT' ? `../HKT.d.ts` : `../lib/${module}.d.ts`,
        sideEffects: false
      },
      null,
      2
    ),
    TE.right
  )
}

const main: Build<void> = pipe(
  copyPackageJson,
  RTE.chain(() => copyFiles),
  RTE.chain(() => makeModules),
)

run(
  main({
    ...fileSystem
  })
)


scripts/FileSystem.ts
import * as TE from 'fp-ts/TaskEither'
import { flow } from 'fp-ts/function'
import * as fs from 'fs'
import G from 'glob'

export interface FileSystem {
  readonly readFile: (path: string) => TE.TaskEither<Error, string>
  readonly writeFile: (path: string, content: string) => TE.TaskEither<Error, void>
  readonly copyFile: (from: string, to: string) => TE.TaskEither<Error, void>
  readonly glob: (pattern: string) => TE.TaskEither<Error, ReadonlyArray<string>>
  readonly mkdir: (path: string) => TE.TaskEither<Error, void>
  readonly moveFile: (from: string, to: string) => TE.TaskEither<Error, void>
}

const readFile = TE.taskify<fs.PathLike, string, NodeJS.ErrnoException, string>(fs.readFile)
const writeFile = TE.taskify<fs.PathLike, string, NodeJS.ErrnoException, void>(fs.writeFile)
const copyFile = TE.taskify<fs.PathLike, fs.PathLike, NodeJS.ErrnoException, void>(fs.copyFile)
const glob = TE.taskify<string, Error, ReadonlyArray<string>>(G)
const mkdirTE = TE.taskify(fs.mkdir)
const moveFile = TE.taskify<fs.PathLike, fs.PathLike, NodeJS.ErrnoException, void>(fs.rename)

export const fileSystem: FileSystem = {
  readFile: (path) => readFile(path, 'utf8'),
  writeFile,
  copyFile,
  glob,
  mkdir: flow(
    mkdirTE,
    TE.map(() => undefined)
  ),
  moveFile
}


src/match.ts
const DEFAULT = '_'
interface None {
  readonly _tag: 'None'
}
interface Some<A> {
  readonly _tag: 'Some'
  readonly value: A
}

export type Option<A> = None | Some<A>
  
interface DefaultCase {
  readonly _tag: typeof DEFAULT
  readonly value: unknown
}
interface DefaultCase {
  readonly [DEFAULT]: () => unknown
}

type _Tag<T extends { readonly _tag: string }> = T['_tag']
type MatchingType<T, Type extends string> = Extract<T, { readonly _tag: Type }>

type X = 'left' | 'right' | 'value'
type Match = {
  readonly [k in X]?: unknown
} & { readonly _tag: string }

type MatchW<A, _Tag extends string, X, R> = (p: Extract<A, { [_tag in _Tag]: X }>) => R
/**
 *  Strict `Pattern Maching` Implementation
 *
 * @example
 * import * as M from 'pattern-matching-ts/lib/match'
 *
 * const optionMatching = M.match<O.Option<string>, string>({
 *   Some: (x) => x.value,
 *   None: () => 'Nothing',
 * })
 *
 * assert.deepStrictEqual(optionMatching(O.some('data')), 'data')
 * assert.deepStrictEqual(optionMatching(O.none), 'Nothing')
 *
 * type Cases = ChangeColor<number> | Move | Write
 *
 * const matchMessage = M.match<Cases, string>({
 *   ChangeColor: ({ value: { r, g, b } }) => `Change the color to Red: ${r} | Green: ${g} | Blue: ${b}`,
 *   Move: ({ value: { x, y } }) => `Move in the x direction: ${x} and in the y direction: ${y}`,
 *   Write: ({ value: { text } }) => `Text message: ${text}`,
 *   _: () => 'Default message'
 * })
 *
 * assert.deepStrictEqual(matchMessage(ChangeColor({ r: 12, g: 20, b: 30 })),'Change the color to Red: 12 | Green: 20 | Blue: 30')
 * assert.deepStrictEqual(matchMessage(Move({ x: 500, y: 100 })),'Move in the x direction: 500 and in the y direction: 100')
 *
 */
export function match<T extends Match | DefaultCase, R = unknown>(
  pattern: T extends Option<unknown>
    ? { [K in _Tag<T>]: (x: MatchingType<T, K>) => R }
    : {
        [K in _Tag<T> | DefaultCase['_tag']]: (x: MatchingType<T, K>) => R
      }
): (x: T) => R {
  return (x) => (pattern as any)[x?._tag ?? DEFAULT](x)
}
/**
 * A Wider pipeable `Pattern Maching` Implementation
 *
 * @example
 * import * as M from 'pattern-matching-ts/lib/match'
 *
 * const optionMatching = (o: O.Option<string>) =>
 *    pipe(
 *      o,
 *      M.matchW('_tag')({
 *       Some: ({ value }) => 'Something: ' + value,
 *       None: () => 'Nothing',
 *     }))
 *
 *   assert.deepStrictEqual(optionMatching(O.some('data')), 'Something: data')
 *   assert.deepStrictEqual(optionMatching(O.none), 'Nothing')
 *
 * const matchResponse = (response: Responses) =>
 *   pipe(
 *     response,
 *     M.matchW('code')({
 *        500: ({ detail }) => ({ message: 'Internal server error', detail }),
 *        404: () => ({ message: 'The page cannot be found!' }),
 *        200: ({ response }) => response.body,
 *        _: () => 'Unexpected response'
 *     })
 *   )
 *
 *   assert.deepStrictEqual(matchResponse({ code: 200, response: { body: ['data'] } }), ['data'])
 *   assert.deepStrictEqual(matchResponse({ code: 500, detail: 'Cannot connect to the database' }), {
 *     message: 'Internal server error',
 *     detail: 'Cannot connect to the database'
 *   })
 *   assert.deepStrictEqual(matchResponse({ code: 404 }), { message: 'The page cannot be found!' })
 * @since 2.0.
 */
export const matchW: <_Tag extends string>(
  _tag: _Tag
) => {
  <
    A extends { [X in _Tag]: string | number | typeof DEFAULT },
    K extends { [X in A[_Tag]]: MatchW<A, _Tag, X, unknown> }
  >(
    k: K
  ): (match: A) => ReturnType<K[keyof K]> | DefaultCase
} = (_tag) => (k: any) => (match: any | DefaultCase) => {
  /* istanbul ignore next */
  return k?.[match?.[_tag]] ? k[match[_tag]](match) : k[DEFAULT](match)
}
